marimbabot_audio/scripts/onset_detection.py

```diff
@@ -0,0 +1,547 @@
+            o
+            for o in onsets_cqt_raw
+            if in_window(o)
+        ]
+
+        winners_idx = []
+        winner_onsets_times = []
+        durations_len = []
+
+        # publish events and plot visualization
+        for o in onsets_cqt:
+            fundamental_frequency, confidence, winner_idx, winner_pm_idx = self.onset_classification(o)
+
+
+            if winner_idx is not None:
+                # find the y-position of onset in spectrum
+                onset_time = self.onsets_to_time_in_spec(o)
+                # extract the duration and loudness of the onset note
+                duration, duration_len, loudness = self.extract_duration(
+                    onset_time=onset_time,
+                    winner_idx=winner_idx,
+                    cqt=cqt,
+                    loudnes_seq=loudness_seq
+                )
+
+                winners_idx.append(winner_idx)
+                winner_onsets_times.append(onset_time)
+                durations_len.append(duration_len)
+
+
+                # publish the onset note to topic
+                t = self.buffer_time + rospy.Duration(o)
+                no = NoteOnset()
+                no.header.stamp = t
+                note = hz_to_note(fundamental_frequency)
+                no.note = note
+                no.confidence = confidence
+                no.duration = duration
+                no.loudness = loudness
+                self.pub_onset.publish(no)
+
+                rospy.loginfo(
+                    f"music note detected: "
+                    f"time:{t} "
+                    f"note:{note} "
+                    f"confidence:{confidence:.4f} "
+                    f"duration:{duration:.4f} "
+                    f"loudness:{loudness:.4f}"
+                )
+
+        # update the sectrum visualization
+        self.update_spectrogram(
+            spec=cqt,
+            ys=winner_onsets_times,
+            xs=winners_idx,
+            durations_len=durations_len,
+        )
+
+        if len(onsets_cqt) == 0:
+            rospy.logdebug("found no onsets")
+        else:
+            rospy.logdebug("found {} onsets".format(len(onsets_cqt)))
+
+        # advance buffer, keep one overlap for next processing
+        self.buffer_time += rospy.Duration(self.window_t)
+        self.buffer = self.buffer[(-2 * self.window_overlap):]
+
+        rospy.loginfo_once("onset detection is online")
+        compute_time = rospy.Time.now() - now
+        self.pub_compute_time.publish(compute_time.to_sec())
+        if compute_time > rospy.Duration(self.window):
+            rospy.logerr("computation took longer than processed window")
+
+    def extract_duration(self, onset_time, winner_idx, cqt, loudnes_seq, frame_len=173, intensity_threshold_ratio=0.4):
+        xs = winner_idx
+        y = onset_time
+
+        max_intensity = np.max(cqt[xs, y - 5:y + 10])
+        loudness = np.max(loudnes_seq[xs, y - 5:y + 10])
+        m = 1
+        while True:
```

Comment:
Maybe this could be done with a for loop:

```python
for(m in range(1, frame_len - y)):
```

Also try to express this without a loop, as it is quite slow in python and most things can be vectorized with numpy.

Reply:
This could look something like this:

```python
offsets = np.arange(1, frame_len - y)
intensities = np.mean(cqt[xs, y + offsets - 1:y + offsets + 1], axis=0)
condition = (intensities / (max_intensity + 0.00001) < intensity_threshold_ratio) | (offsets + y >= frame_len - 1)
m = np.argmax(condition)
```

Disclaimer: I have not checked this code, and it is only a rough idea, but by using broadcasting something like this could be achieved.

Reply:
later on when speed not satisfied.